Upgrade to Rails 3

tylerritchie opened the pull request:
Howdy @Nonsense331 and @samsonasu (cc: @hadley) I know this is sort of a drive-by PR but I had started down the path of upgrading hadley/crantastic master before I saw this fork had activity, so I switched gears and got this all updated.

I mostly fixed tests, but I also made the session secret an ENV value and pinned authlogic to sha512, and cleaned up some gems.

I'm pretty sure there are all kinds of things that are still broken, but a green test suite is a much nicer place to start from.

tylerritchie commented:
@Nonsense331 gentlest of bumps, if you're interested I can resolve these merge conflicts so you can merge this in.

samsonasu commented on config/initializers/secret_token.rb:
```diff
@@ -4,4 +4,8 @@
 # Make sure the secret is at least 30 characters and all random,
 # no regular words or you'll be exposed to dictionary attacks.
-Crantastic::Application.config.secret_token = '52b1567b80b6c472cd1348eb23da1512c9adae4ce374e5fb48965983cf7708b2e2e972d9c347df91a44bdda6a22b4f79ca9b8e63c9c27d7a02a3b61037a5f7f4'
+Crantastic::Application.config.secret_token = if Rails.env.development? or Rails.env.test?
```
This should just be `ENV['SECRET_TOKEN']`always, it's not a big deal to put this in local configuration.  Also we should consider adding [Figaro](https://github.com/laserlemon/figaro) if it's not already present instead of that 00-load-config initializer

boatrite replied:
Also, `or` and `and` are not the same as `||` and `&&`. The English words are for control flow, not boolean logic.

samsonasu commented on spec/features/auth_feature_spec.rb:
```diff
@@ -9,7 +10,8 @@
     fill_in "Confirm password", :with => "test"
     check "user_tos"
-    UserMailer.should_receive(:deliver_activation_instructions)
+    UserMailer.should_receive(:activation_instructions).and_return(email)
```
I can never keep up with what version of rspec we're on but we should use the `allow(UserMailer).to_receive` syntax instead if it's available

boatrite replied:
I noticed it's version 2.99. It's the version that deprecates the old syntax but brings in the new one to make converting easier.  In fact, @tylerritchie, if the tests are passing, just run [transpec](https://github.com/yujinakayama/transpec) on it and it will convert mostly everything for you.

samsonasu commented on spec/models/package_spec.rb:
```diff
@@ -53,28 +74,28 @@
     prev_time = pkg.updated_at
     create :version, :package => pkg, :maintainer => Author.first, :version => "5.3"
-    (pkg.updated_at > prev_time).should be_true
+    (pkg.updated_at > prev_time).should be_truthy
```
shouldn't this actually return true not just truthy?

boatrite replied:
Agreed. `expect(stuff).to eq true`

samsonasu commented on spec/models/package_user_spec.rb:
```diff
@@ -1,3 +1,19 @@
+# Indexes
+#
+#  index_package_user_on_package_id_and_user_id  (package_id,user_id) UNIQUE
+#
```
We usually don't include these headers because of the risk of them becoming outdated.  I can live with them in the models but I think it's overkill to have them in specs too, too much duplication imo

samsonasu commented on spec/factories/user_factory.rb:
```diff
@@ -1,7 +1,42 @@
   factory :user do
     email { Faker::Internet.email }
-    login { Faker::Name.first_name }
+    login { Faker::Internet.user_name(3..100) }
```
We should convert this and also email to use factorygirl sequences -- it's rare to get duplicates but it does happen, i..e 

```FactoryGirl.define do
  factory :user do
    sequence(:email) { |n| "email#{n}@example.com" }
  end
end```

samsonasu commented:
Thanks for the PR, and thanks for the postcard :) We're looking at this today and in the meantime can you please rebase to the current rails-3? 

We will also probably marge rails-3 to master this month since that branch has been in production for quite a while now and theres no chance we'll go back to rails 2.

Nonsense331 commented on app/mailers/digest_mailer.rb:
```diff
@@ -0,0 +1,11 @@
+
+  def weekly_digest(digest)
+    @digest = digest
+    mail to: "r-help@r-project.org",
```
Is this a real email?

hadley replied:
It is, but we shouldn't send anything to it without some review

tylerritchie replied:
I just moved the mailer, I didn't change the rake task that sends it (Which I don't think has been run since October?).

boatrite commented on app/models/author.rb:
```diff
@@ -1,3 +1,14 @@
+# == Schema Information
```
Let's aggressively delete these. They go out of date too easily and it's just as easy to pop open schema.rb for the source of truth.

boatrite commented on config/initializers/00-load_config.rb:
```diff
@@ -16,6 +16,7 @@
+  ENV['SECRET_TOKEN'] = private_config["SECRET_TOKEN"]
   ENV['TWITTER_OAUTH_TOKEN_SECRET'] = private_config["TWITTER_OAUTH_TOKEN_SECRET"]
   ENV['TWITTER_OAUTH_TOKEN'] = private_config["TWITTER_OAUTH_TOKEN"]
   ENV['TWITTER_CONSUMER_KEY'] = private_config["TWITTER_CONSUMER_KEY"]
```
Lets replace this file with [Figaro](https://github.com/laserlemon/figaro). Add config/application.yml, add it to gitignore, and move all this to that file (and this private_config stuff)

boatrite commented on spec/controllers/packages_controller_spec.rb:
```diff
@@ -13,66 +13,67 @@
-    Package.should_receive(:find_by_param).with("aaMI").and_return(pkg_mock)
+        package = FactoryGirl.create(:package, :name => 'aaMI')
+        controller.instance_eval do
+          current_user.should_receive(:toggle_usage).with(package).and_return(true)
```
The new syntax is `allow(current_user).to receive(:toggle_user).with(package).and_return(true)`

boatrite commented on spec/controllers/reviews_controller_spec.rb:
```diff
@@ -47,12 +47,12 @@
     get :edit, :id => 1
     response.should_not be_success
-    response.status.should == "403 Forbidden"
+    response.status.should == 403
```
Use expect syntax

boatrite commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -31,14 +31,13 @@
     login_as_user
     get :new
-    response.should_not render_template
+    response.should_not render_template(:new)
```
expect syntax

boatrite commented on spec/controllers/votes_controller_spec.rb:
```diff
@@ -20,8 +20,8 @@
-    @user.uses?(ggplot).should be_true
-    @user.uses?(rjson).should be_true
+    @user.uses?(ggplot).should be_truthy
+    @user.uses?(rjson).should be_truthy
```
`expect(@user.uses? rjson).to eq true`.

boatrite commented on spec/factories/package_factory.rb:
```diff
@@ -1,5 +1,29 @@
+# == Schema Information
```
These don't need to be in factories either

boatrite commented on spec/factories/tagging_factory.rb:
```diff
@@ -1,8 +1,29 @@
     trait :with_package do
       package
     end
+    trait :with_tag do
```
line breaks between trait blocks

boatrite commented on spec/mailers/user_mailer_spec.rb:
```diff
@@ -0,0 +1,19 @@
+      email.to.should == [user.email]
+    end
+
+    it "should include an activation link" do
```
Remove 'should' from descriptions. `it "includes an activation link"`

boatrite commented on spec/spec_helper.rb:
```diff
@@ -2,7 +2,7 @@
 require File.expand_path("../../config/environment", __FILE__)
```
After you update to rspec 3, backup this file and regenerate via `rails g rspec:install`. RSpec 3 adds in spec_helper *and* rails_helper.

Remove the giant block comment from spec_helper. It has sensible defaults (mostly) such as enforcing the new syntax (which transpec can convert all the files to)

Move anything extra that was added in this spec file to its own file in spec/support. e.g. Since we're using capybara, create spec/support/capybara.rb and include pertinent code there. I like to keep the generated helper files at their defaults as it makes understanding what we've changed easier (it's only the stuff in spec/support).

You'll have to uncomment the line in rails_helper that includes all the files in spec/support

boatrite commented on spec/spec_helper.rb:
```diff
@@ -70,5 +70,6 @@ def self.execute(request)
   config.infer_spec_type_from_file_location!
 
   include FactoryGirl::Syntax::Methods
```
We can just delete this. We always call FactoryGirl methods with FactoryGirl.whatever anyways

boatrite commented on spec/spec_helper.rb:
```diff
@@ -70,5 +70,6 @@ def self.execute(request)
 
   include FactoryGirl::Syntax::Methods
+  include Authlogic::TestCase
   config.include FeatureHelpers, type: :feature
```
spec/support/feature_helpers.rb

boatrite commented on spec/views/authors/index.html.erb_spec.rb:
```diff
@@ -1,14 +1,11 @@
-require File.expand_path(File.dirname(__FILE__) + '/../../spec_helper')
+require 'spec_helper'
```
All of these will change to `require 'rails_helper'`

boatrite commented:
Great work. This was a huge PR but it looks like it does a lot of good things.

tylerritchie commented:
@boatrite @Nonsense331 @samsonasu @hadley I removed the authlogic sha512 commit I mentioned after the mention that the rails-3 branch has been operating in production. And squashed the first pass of getting the specs working.

I also removed the annotate gem and annotations, added figaro, and upgraded to rspec 3.

All tests are passing.

samsonasu commented:
@tylerritchie thanks so much for the help!
